artifactory_cleanup/rules/delete.py

```diff
@@ -92,37 +93,85 @@ class delete_empty_folder(Rule):
+        aql_query_list.append(all_files_dict)
         return aql_query_list
 
     def _filter_result(self, result_artifact):
-        r = self.artifactory_session.get(
-            "{}/api/repositories?type=local".format(self.artifactory_server)
-        )
-        r.raise_for_status()
-        repositories = r.json()
-
-        for count, repository in enumerate(repositories, start=1):
-            if repository["packageType"] == "GitLfs":
-                # GitLfs should be handled by the jfrog cli: https://jfrog.com/blog/clean-up-your-git-lfs-repositories-with-jfrog-cli/
-                print(
-                    f"Skipping '{repository['key']}' because it is a Git LFS repository"
-                )
-                continue
-
-            url = "{}/api/plugins/execute/deleteEmptyDirsPlugin?params=paths={}".format(
-                self.artifactory_server, repository["key"]
-            )
-
-            print(
-                f"Deleting empty folders for '{repository['key']}' - {count} of {len(repositories)}"
-            )
-            r = self.artifactory_session.post(url)
-            r.raise_for_status()
-
-        return []
+
+        # convert list of files to dict
+        # Source: https://stackoverflow.com/a/58917078
+
+        def nested_dict():
+            """
+            Creates a default dictionary where each value is an other default dictionary.
+            """
+            return defaultdict(nested_dict)
+
+        def default_to_regular(d):
+            """
+            Converts defaultdicts of defaultdicts to dict of dicts.
+            """
+            if isinstance(d, defaultdict):
+                d = {k: default_to_regular(v) for k, v in d.items()}
+            return d
+
+        def get_path_dict(artifacts):
+            new_path_dict = nested_dict()
+            for artifact in artifacts:
+                parts = artifact["path"].split("/")
+                if parts:
+                    marcher = new_path_dict
+                    for key in parts:
+                        # We need the repo for the root level folders. They are not in the
+                        # artifacts list
+                        marcher[key]["data"] = {"repo": artifact["repo"]}
+                        marcher = marcher[key]["children"]
+                    marcher[artifact["name"]]["data"] = artifact
+            return default_to_regular(new_path_dict)
+
+        artifact_tree = get_path_dict(result_artifact)
+
+        # Artifactory also returns the directory itself. We need to remove it from the list
+        # since that tree branch has no children assigned
+        if "." in artifact_tree:
+            del artifact_tree["."]
+
+        # Now we have a dict with all folders and files
+        # An empty folder is represented if it is a dict and does not have any keys
+
+        def get_folder_artifacts_with_no_children(item, path=""):
+
+            empty_folder_artifacts = deque()
+
+            def _add_to_del_list(key):
```

Comment:
Let's move the function outside `_add_to_del_list` too

Reply:
I will also refactor this and move it to utils.py